File: dnn-providers/hip-kernel-provider/rocKE/Python/rocke/helpers/io.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""Dtype-dispatched I/O helpers.

This module deduplicates the small ``_io_type`` / ``_load_vec`` /
``_store_vec`` / ``_global_load_scalar`` blocks that every small-op
file (elementwise, layernorm, rmsnorm, reduce, transpose) was copying
verbatim. The helpers are dtype-string-tolerant: ``"f16"``, ``"fp16"``,
and ``"bf16"`` all resolve to the canonical IR type, so ports of CK
Tile reference kernels (which use ``"fp16"``) don't have to translate.

The convention matches CK Tile's ``type_convert<DstT, SrcT>`` family in
``include/ck_tile/core/numeric/type_convert.hpp``: I/O is always at the
native storage dtype (f16/bf16); compute is in f32 (the
``ComputeDataType`` alias in every CK Tile op's TypeConfig). The
:func:`load_vec_as_f32` / :func:`load_scalar_as_f32` convenience wrap
the most common dtype-promote-on-read pattern.
"""

from __future__ import annotations

from typing import Literal

from ..core.ir import BF16, F16, IRBuilder, Type, Value


__all__ = [
    "DType",
    "io_ir_type",
    "load_lane_slice_f32",
    "load_scalar",
    "load_scalar_as_f32",
    "load_vec",
    "load_vec_as_f32",
    "pack_f32_to",
    "pack_quant_chunk_f32",
    "store_packed_chunk",
    "store_scalar",
    "store_scalar_from_f32",
    "store_vec",
    "vector_row_copy",
]


# Aliases the small ops use. ``"fp16"`` is the CK Tile / ``UniversalGemm``
# spelling; we accept it to make port snippets read literally.
DType = Literal["f16", "fp16", "bf16"]


def io_ir_type(dtype: str) -> Type:
    """Map a dtype string to the canonical IR type object.

    Accepts ``"f16"``, ``"fp16"`` (alias), or ``"bf16"``. Raises
    :class:`ValueError` for any other value -- f8 and i8 paths go
    through their own helpers because their compute dtype isn't f32.
    """
    if dtype in ("f16", "fp16"):
        return F16
    if dtype == "bf16":
        return BF16
    raise ValueError(f"unsupported I/O dtype {dtype!r}; expected f16/fp16/bf16")


def load_scalar(b: IRBuilder, ptr: Value, idx: Value, *, dtype: str) -> Value:
    """One scalar global load. Returns a value in the native dtype.

    Caller is responsible for any subsequent ``cast_to_f32`` if compute
    needs f32. Use :func:`load_scalar_as_f32` for the common
    "load + promote" pattern.
    """
    if dtype in ("f16", "fp16"):
        return b.global_load_f16(ptr, idx)
    if dtype == "bf16":
        return b.global_load_bf16(ptr, idx)
    raise ValueError(f"unsupported I/O dtype {dtype!r}")


def load_scalar_as_f32(b: IRBuilder, ptr: Value, idx: Value, *, dtype: str) -> Value:
    """One scalar global load promoted to f32.

    Equivalent to ``cast_to_f32(load_scalar(...))``; the helper exists
    because every norm / reduce / elementwise tail path does exactly
    this two-step.
    """
    return b.cast_to_f32(load_scalar(b, ptr, idx, dtype=dtype))


def load_vec(b: IRBuilder, ptr: Value, idx: Value, *, dtype: str, n: int) -> Value:
    """Vectorised global load of ``n`` consecutive elements.

    Supports ``n in {2, 4, 8}`` for f16/bf16; ``n=1`` is rejected
    because the IR distinguishes scalar vs vector loads and the n=1
    case would silently lose vectorisation. Use :func:`load_scalar`
    when a scalar is what you want.
    """
    if n not in (2, 4, 8):
        raise ValueError(f"load_vec n must be 2/4/8 (got {n}); use load_scalar for n=1")
    ty = io_ir_type(dtype)
    if dtype in ("f16", "fp16"):
        return b.global_load_vN_f16(ptr, idx, n)
    return b.global_load_vN(ptr, idx, ty, n)


def load_vec_as_f32(
    b: IRBuilder, ptr: Value, idx: Value, *, dtype: str, n: int
) -> list[Value]:
    """Vectorised load + per-lane f32 promotion.

    Returns a list of ``n`` scalar f32 :class:`Value`\\s, one per element
    of the loaded vector. This is the canonical "ingest into f32 compute
    registers" pattern used by every norm/reduce kernel.
    """
    v = load_vec(b, ptr, idx, dtype=dtype, n=n)
    return [b.cast_to_f32(b.vec_extract(v, i)) for i in range(n)]


# Power-of-two vector widths the DSL's ``global_load_vN`` covers. EPT
# values outside this set (1 / 3 for head_size 64 / 192) fall back to
# per-element scalar loads so every supported head size keeps working.
_VEC_WIDTHS = (2, 4, 8)


def load_lane_slice_f32(
    b: IRBuilder,
    ptr: Value,
    row_base: Value,
    lane_d_base: Value,
    *,
    dtype: str,
    ept: int,
) -> list[Value]:
    """Load this lane's ``EPT`` consecutive elements as a list of f32.

    Uses one vectorised ``global_load_vN`` + ``vec_extract`` chain when
    ``EPT`` is a supported vector width (2 / 4 / 8 -- one VMEM
    transaction per call). Falls back to per-element scalar loads for
    the ``EPT == 1`` (head_size=64) and ``EPT == 3`` (head_size=192,
    not a power of two) corner cases so every supported head size keeps
    working.

    Promoted from the AST-identical local ``_load_lane_slice_f32`` in
    :mod:`rocke.instances.common.fmha_bwd` and
    :mod:`rocke.instances.common.fmha_splitkv_decode`. Matches the
    per-warp K/V/Q load pattern used by CK Tile's ``BlockFmhaBwd*``
    register-tile loads (``load_tile`` over a distributed
    ``rt<bf16, ..., row_l, rt_16x32_s>`` tensor) and AITER's varlen bwd
    ``tl.load(ptr + offs, ...)`` 8-element vector loads.
    """
    if ept in _VEC_WIDTHS:
        return load_vec_as_f32(b, ptr, b.add(row_base, lane_d_base), dtype=dtype, n=ept)
    return [
        load_scalar_as_f32(
            b,
            ptr,
            b.add(row_base, b.add(lane_d_base, b.const_i32(k))),
            dtype=dtype,
        )
        for k in range(ept)
    ]


def store_scalar(
    b: IRBuilder, ptr: Value, idx: Value, value: Value, *, dtype: str
) -> None:
    """One scalar global store.

    ``value`` must be in the native dtype already. Use
    :func:`store_scalar_from_f32` to handle the f32 -> {f16, bf16}
    cast for you.
    """
    if dtype not in ("f16", "fp16", "bf16"):
        raise ValueError(f"unsupported I/O dtype {dtype!r}")
    b.global_store(ptr, idx, value)


def store_scalar_from_f32(
    b: IRBuilder, ptr: Value, idx: Value, value_f32: Value, *, dtype: str
) -> None:
    """Trunc an f32 value to ``dtype`` and store it scalar.

    Most norm/reduce kernels keep the per-element accumulator in f32
    and want to fuse the trunc + store. This is the helper for that
    last mile.
    """
    target = io_ir_type(dtype)
    b.global_store(ptr, idx, b.cast_f32_to(value_f32, target))


def store_vec(b: IRBuilder, ptr: Value, idx: Value, value: Value, *, n: int) -> None:
    """Vectorised global store. ``value`` must already be a ``<n x T>``
    vector in the target dtype (use :func:`pack_f32_to` to assemble
    one from a list of f32 scalars).
    """
    if n not in (2, 4, 8):
        raise ValueError(
            f"store_vec n must be 2/4/8 (got {n}); use store_scalar for n=1"
        )
    b.global_store_vN(ptr, idx, value, n)


def pack_f32_to(b: IRBuilder, scalars_f32: list[Value], *, dtype: str) -> Value:
    """Trunc a list of f32 scalars to ``dtype`` and pack into a vector.

    The dual of :func:`load_vec_as_f32`: every norm / elementwise
    kernel finishes with this exact pattern before writing back to
    global. Returns a ``<len(scalars_f32) x dtype>`` vector.
    """
    target = io_ir_type(dtype)
    casts = [b.cast_f32_to(v, target) for v in scalars_f32]
    return b.vec_pack(casts, target)


def vector_row_copy(
    b: IRBuilder,
    *,
    src: Value,
    dst: Value,
    src_base: Value,
    dst_base: Value,
    H: int,
    dtype: str,
    vec_bytes: int = 16,
) -> None:
    """Vectorised row copy along a head / hidden dimension.

    Promotes the inline ``_copy_row_vec`` from
    :mod:`rocke.instances.common.fmha_appendkv` into a shared helper so the
    same 16-byte-vector pattern can be reused by ``moe_gather`` and
    ``fmha_bwd`` postlude.

    The copy issues ``H // (vec_bytes / sizeof(elem))`` aligned vmem
    transactions plus a scalar tail for any residual elements; the
    AMDGPU backend coalesces the vector form into a single
    ``buffer_load_dwordx4`` + ``buffer_store_dwordx4`` pair when
    pointer alignment supports it. Matches CK Tile's
    ``store_tile(k_dram_block_window, knew_tile)`` semantics in
    ``include/ck_tile/ops/fmha/pipeline/block_fmha_fwd_appendkv_pipeline.hpp``.
    """
    ty = io_ir_type(dtype)
    elem_bytes = 2  # f16 / bf16 storage width
    vec = vec_bytes // elem_bytes
    if vec not in (2, 4, 8):
        raise ValueError(
            f"vector_row_copy: vec_bytes {vec_bytes} maps to vec={vec}; "
            "expected 4/8/16-byte aligned"
        )
    n_chunks = H // vec
    for c in range(n_chunks):
        d = c * vec
        src_addr = b.add(src_base, b.const_i32(d))
        dst_addr = b.add(dst_base, b.const_i32(d))
        v = b.global_load_vN(src, src_addr, ty, vec, align=vec * elem_bytes)
        b.global_store_vN(dst, dst_addr, v, vec, align=vec * elem_bytes)
    for d in range(n_chunks * vec, H):
        s = load_scalar_as_f32(b, src, b.add(src_base, b.const_i32(d)), dtype=dtype)
        store_scalar_from_f32(b, dst, b.add(dst_base, b.const_i32(d)), s, dtype=dtype)


def pack_quant_chunk_f32(
    b: IRBuilder,
    chunk_f32: list[Value],
    *,
    qdtype: str,
) -> Value:
    """Pack a list of 4 f32 scalars into a single packed quant vector.

    Promotes the inline ``_pack_quant_chunk_f32`` from
    :mod:`rocke.instances.common.smoothquant` (and the duplicate copy in
    :mod:`rocke.instances.common.moe_smoothquant`) into a shared helper.

    Routes through the matching packed cvt primitive on the IR:

    * ``"fp8e4m3"`` / ``"fp8"`` — :meth:`IRBuilder.cvt_pk_fp8_f32x4`
    * ``"bf8e5m2"`` / ``"bf8"`` — :meth:`IRBuilder.cvt_pk_bf8_f32x4`
    * ``"i8"`` — :meth:`IRBuilder.cvt_pk_i8_f32x4` (P09)

    The caller still owns the scale-multiply + clamp prelude (see
    :func:`rocke.helpers.quant.quantize_scalar_f32`); this helper
    only does the packed cvt over a chunk of 4 f32 scalars.
    """
    if len(chunk_f32) != 4:
        raise ValueError(
            f"pack_quant_chunk_f32 expects exactly 4 f32 scalars, got {len(chunk_f32)}"
        )
    from ..core.ir import F32

    vec_in = b.vec_pack(chunk_f32, F32)
    if qdtype in ("fp8", "fp8e4m3"):
        return b.cvt_pk_fp8_f32x4(vec_in)
    if qdtype in ("bf8", "bf8e5m2"):
        return b.cvt_pk_bf8_f32x4(vec_in)
    if qdtype == "i8":
        return b.cvt_pk_i8_f32x4(vec_in)
    raise ValueError(
        f"pack_quant_chunk_f32: unsupported qdtype {qdtype!r}; expected fp8 / bf8 / i8"
    )


def store_packed_chunk(
    b: IRBuilder,
    *,
    dst: Value,
    dst_base: Value,
    packed_vec: Value,
) -> None:
    """Store a 4-byte packed quant vector to global memory.

    Companion to :func:`pack_quant_chunk_f32`: takes the ``<4 x i8>``
    /``<4 x fp8e4m3>`` / ``<4 x bf8e5m2>`` produced by the cvt and
    emits a single 4-byte global store. AMDGPU's backend coalesces
    adjacent lanes' chunks into one ``global_store_dwordx<N>``
    transaction when alignments line up.
    """
    b.global_store_vN(dst, dst_base, packed_vec, 4, align=4)

File: dnn-providers/hip-kernel-provider/rocke/platform/Cpp/helpers/io.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/*
 * helper_rocke.helpers.io.c -- C99 port of rocke.helpers.io.io_ir_type,
 * store_scalar_from_f32, load_scalar, load_scalar_as_f32, load_vec,
 * load_vec_as_f32, load_lane_slice_f32, store_vec, pack_f32_to and
 * vector_row_copy.
 *
 * Faithful translation of:
 *
 *     def io_ir_type(dtype: str) -> Type:
 *         if dtype in ("f16", "fp16"): return F16
 *         if dtype == "bf16":          return BF16
 *         raise ValueError(
 *             f"unsupported I/O dtype {dtype!r}; expected f16/fp16/bf16")
 *
 * Mapping invariants (must stay byte-identical to the Python so downstream IR is
 * identical):
 *   "f16"  -> rocke_f16()    (Python F16)
 *   "fp16" -> rocke_f16()    (alias; Python F16)
 *   "bf16" -> rocke_bf16()   (Python BF16)
 *   else   -> NULL         (Python ValueError)
 */

#include "rocke/helper_rocke.helpers.io.h"

#include <string.h>

#include "rocke/arena.h" /* rocke_arena_alloc */
#include "rocke/ir_internal.h" /* rocke_i_set_err, rocke_i_live */

/* Power-of-two vector widths the DSL's global_load_vN covers. Mirrors the
 * module-level _VEC_WIDTHS = (2, 4, 8) in rocke.helpers.io. */
static int rocke_io_is_vec_width(int n)
{
    return n == 2 || n == 4 || n == 8;
}

const rocke_type_t* rocke_io_ir_type(const char* dtype)
{
    if(dtype == NULL)
    {
        return NULL;
    }
    /* `dtype in ("f16", "fp16")` -> F16 */
    if(strcmp(dtype, "f16") == 0 || strcmp(dtype, "fp16") == 0)
    {
        return rocke_f16();
    }
    /* `dtype == "bf16"` -> BF16 */
    if(strcmp(dtype, "bf16") == 0)
    {
        return rocke_bf16();
    }
    /* Python: raise ValueError. No builder here, so signal via NULL. */
    return NULL;
}

const rocke_type_t* rocke_b_io_ir_type(rocke_ir_builder_t* b, const char* dtype)
{
    const rocke_type_t* ty;

    /* Sticky-error model: a failed builder makes every call a NULL no-op. */
    if(!rocke_i_live(b))
    {
        return NULL;
    }

    ty = rocke_io_ir_type(dtype);
    if(ty == NULL)
    {
        /* Mirror the Python ValueError, including the {dtype!r} single-quote
         * repr for the (non-NULL) string case. NULL is reported as "None" to
         * match Python's repr(None). */
        return (const rocke_type_t*)rocke_i_set_err(
            b,
            ROCKE_ERR_VALUE,
            "unsupported I/O dtype %s%s%s; expected f16/fp16/bf16",
            dtype ? "'" : "",
            dtype ? dtype : "None",
            dtype ? "'" : "");
    }
    return ty;
}

void rocke_b_store_scalar_from_f32(rocke_ir_builder_t* b,
                                   rocke_value_t* ptr,
                                   rocke_value_t* idx,
                                   rocke_value_t* value_f32,
                                   const char* dtype)
{
    const rocke_type_t* target;
    rocke_value_t* cast;

    /* Sticky-error model: a failed builder makes every call a NULL no-op. */
    if(!rocke_i_live(b))
    {
        return;
    }

    /* Python: target = io_ir_type(dtype). Use the builder-aware variant so an
     * unsupported dtype records the same ValueError on the sticky-error model;
     * on failure it returns NULL and we stop (Python would have raised). */
    target = rocke_b_io_ir_type(b, dtype);
    if(target == NULL)
    {
        return;
    }

    /* Python: b.global_store(ptr, idx, b.cast_f32_to(value_f32, target)).
     * align defaults to 0 (Python global_store takes no align kwarg here). */
    cast = rocke_b_cast_f32_to(b, value_f32, target);
    rocke_b_global_store(b, ptr, idx, cast, 0);
}

rocke_value_t* rocke_b_load_scalar(rocke_ir_builder_t* b,
                                   rocke_value_t* ptr,
                                   rocke_value_t* idx,
                                   const char* dtype)
{
    if(!rocke_i_live(b))
    {
        return NULL;
    }

    /* Python: if dtype in ("f16", "fp16"): return b.global_load_f16(ptr, idx) */
    if(dtype != NULL && (strcmp(dtype, "f16") == 0 || strcmp(dtype, "fp16") == 0))
    {
        /* Python passes no align kwarg -> default. 0 selects the builder default. */
        return rocke_b_global_load_f16(b, ptr, idx, 0);
    }
    /* Python: if dtype == "bf16": return b.global_load_bf16(ptr, idx) */
    if(dtype != NULL && strcmp(dtype, "bf16") == 0)
    {
        return rocke_b_global_load_bf16(b, ptr, idx, 0);
    }
    /* Python: raise ValueError(f"unsupported I/O dtype {dtype!r}"). Note this
     * message has NO ", expected ..." suffix -- it is load_scalar's own text,
     * distinct from io_ir_type's. Mirror it exactly. */
    return (rocke_value_t*)rocke_i_set_err(b,
                                           ROCKE_ERR_VALUE,
                                           "unsupported I/O dtype %s%s%s",
                                           dtype ? "'" : "",
                                           dtype ? dtype : "None",
                                           dtype ? "'" : "");
}

rocke_value_t* rocke_b_load_scalar_as_f32(rocke_ir_builder_t* b,
                                          rocke_value_t* ptr,
                                          rocke_value_t* idx,
                                          const char* dtype)
{
    rocke_value_t* v;

    if(!rocke_i_live(b))
    {
        return NULL;
    }

    /* Python: return b.cast_to_f32(load_scalar(b, ptr, idx, dtype=dtype)).
     * load_scalar records the ValueError on an unsupported dtype and returns
     * NULL; cast_to_f32 of NULL on an errored builder is itself a NULL no-op. */
    v = rocke_b_load_scalar(b, ptr, idx, dtype);
    if(v == NULL)
    {
        return NULL;
    }
    return rocke_b_cast_to_f32(b, v);
}

rocke_value_t* rocke_b_load_vec(
    rocke_ir_builder_t* b, rocke_value_t* ptr, rocke_value_t* idx, const char* dtype, int n)
{
    const rocke_type_t* ty;

    if(!rocke_i_live(b))
    {
        return NULL;
    }

    /* Python checks n FIRST: if n not in (2,4,8): raise ValueError(...). */
    if(!rocke_io_is_vec_width(n))
    {
        return (rocke_value_t*)rocke_i_set_err(
            b, ROCKE_ERR_VALUE, "load_vec n must be 2/4/8 (got %d); use load_scalar for n=1", n);
    }
    /* Python: ty = io_ir_type(dtype). Builder-aware: an unsupported dtype
     * records io_ir_type's ValueError and returns NULL here. */
    ty = rocke_b_io_ir_type(b, dtype);
    if(ty == NULL)
    {
        return NULL;
    }
    /* Python: if dtype in ("f16","fp16"): return b.global_load_vN_f16(ptr,idx,n)
     *         return b.global_load_vN(ptr, idx, ty, n)
     * Python passes no align -> default (0). */
    if(strcmp(dtype, "f16") == 0 || strcmp(dtype, "fp16") == 0)
    {
        return rocke_b_global_load_vN_f16(b, ptr, idx, n, 0);
    }
    return rocke_b_global_load_vN(b, ptr, idx, ty, n, 0);
}

int rocke_b_load_vec_as_f32(rocke_ir_builder_t* b,
                            rocke_value_t* ptr,
                            rocke_value_t* idx,
                            const char* dtype,
                            int n,
                            rocke_value_t** out)
{
    rocke_value_t* v;
    int i;

    if(!rocke_i_live(b))
    {
        return 0;
    }
    if(out == NULL)
    {
        return 0;
    }

    /* Python: v = load_vec(b, ptr, idx, dtype=dtype, n=n) */
    v = rocke_b_load_vec(b, ptr, idx, dtype, n);
    if(v == NULL)
    {
        return 0;
    }
    /* Python: return [b.cast_to_f32(b.vec_extract(v, i)) for i in range(n)] */
    for(i = 0; i < n; ++i)
    {
        out[i] = rocke_b_cast_to_f32(b, rocke_b_vec_extract(b, v, i));
    }
    return 1;
}

int rocke_b_load_lane_slice_f32(rocke_ir_builder_t* b,
                                rocke_value_t* ptr,
                                rocke_value_t* row_base,
                                rocke_value_t* lane_d_base,
                                const char* dtype,
                                int ept,
                                rocke_value_t** out)
{
    int k;

    if(!rocke_i_live(b))
    {
        return 0;
    }
    if(out == NULL)
    {
        return 0;
    }

    /* Python: if ept in _VEC_WIDTHS:
     *             return load_vec_as_f32(
     *                 b, ptr, b.add(row_base, lane_d_base), dtype=dtype, n=ept) */
    if(rocke_io_is_vec_width(ept))
    {
        return rocke_b_load_vec_as_f32(
            b, ptr, rocke_b_add(b, row_base, lane_d_base), dtype, ept, out);
    }
    /* Python scalar fallback:
     *   return [load_scalar_as_f32(
     *               b, ptr,
     *               b.add(row_base, b.add(lane_d_base, b.const_i32(k))),
     *               dtype=dtype)
     *           for k in range(ept)] */
    for(k = 0; k < ept; ++k)
    {
        rocke_value_t* addr
            = rocke_b_add(b, row_base, rocke_b_add(b, lane_d_base, rocke_b_const_i32(b, k)));
        out[k] = rocke_b_load_scalar_as_f32(b, ptr, addr, dtype);
        if(out[k] == NULL)
        {
            /* load_scalar_as_f32 recorded the ValueError already. */
            return 0;
        }
    }
    return 1;
}

void rocke_b_store_vec(
    rocke_ir_builder_t* b, rocke_value_t* ptr, rocke_value_t* idx, rocke_value_t* value, int n)
{
    if(!rocke_i_live(b))
    {
        return;
    }

    /* Python: if n not in (2,4,8): raise ValueError(...) */
    if(!rocke_io_is_vec_width(n))
    {
        (void)rocke_i_set_err(
            b, ROCKE_ERR_VALUE, "store_vec n must be 2/4/8 (got %d); use store_scalar for n=1", n);
        return;
    }
    /* Python: b.global_store_vN(ptr, idx, value, n). No align -> default (0). */
    rocke_b_global_store_vN(b, ptr, idx, value, n, 0);
}

rocke_value_t* rocke_b_pack_f32_to(rocke_ir_builder_t* b,
                                   rocke_value_t* const* scalars_f32,
                                   int n,
                                   const char* dtype)
{
    const rocke_type_t* target;
    rocke_value_t** casts;
    int i;

    if(!rocke_i_live(b))
    {
        return NULL;
    }

    /* Python: target = io_ir_type(dtype). Builder-aware -> records ValueError
     * on an unsupported dtype and returns NULL. */
    target = rocke_b_io_ir_type(b, dtype);
    if(target == NULL)
    {
        return NULL;
    }
    /* Python: casts = [b.cast_f32_to(v, target) for v in scalars_f32].
     * Build the cast list in the arena so vec_pack gets a contiguous array. */
    casts = (rocke_value_t**)rocke_arena_alloc(&b->arena, (size_t)n * sizeof(*casts));
    if(casts == NULL)
    {
        return NULL;
    }
    for(i = 0; i < n; ++i)
    {
        casts[i] = rocke_b_cast_f32_to(b, scalars_f32[i], target);
    }
    /* Python: return b.vec_pack(casts, target). */
    return rocke_b_vec_pack(b, casts, n, target);
}

void rocke_b_vector_row_copy(rocke_ir_builder_t* b,
                             rocke_value_t* src,
                             rocke_value_t* dst,
                             rocke_value_t* src_base,
                             rocke_value_t* dst_base,
                             int H,
                             const char* dtype,
                             int vec_bytes)
{
    const rocke_type_t* ty;
    const int elem_bytes = 2; /* f16 / bf16 storage width */
    int vec;
    int n_chunks;
    int c;
    int d;

    if(!rocke_i_live(b))
    {
        return;
    }

    /* Python: ty = io_ir_type(dtype). Builder-aware -> records ValueError on an
     * unsupported dtype and returns NULL. */
    ty = rocke_b_io_ir_type(b, dtype);
    if(ty == NULL)
    {
        return;
    }
    /* Python: vec = vec_bytes // elem_bytes; if vec not in (2,4,8): raise. */
    vec = vec_bytes / elem_bytes;
    if(!rocke_io_is_vec_width(vec))
    {
        (void)rocke_i_set_err(b,
                              ROCKE_ERR_VALUE,
                              "vector_row_copy: vec_bytes %d maps to vec=%d; expected 4/8/16-byte "
                              "aligned",
                              vec_bytes,
                              vec);
        return;
    }
    /* Python: n_chunks = H // vec
     *         for c in range(n_chunks):
     *             d = c * vec
     *             src_addr = b.add(src_base, b.const_i32(d))
     *             dst_addr = b.add(dst_base, b.const_i32(d))
     *             v = b.global_load_vN(src, src_addr, ty, vec, align=vec*elem_bytes)
     *             b.global_store_vN(dst, dst_addr, v, vec, align=vec*elem_bytes) */
    n_chunks = H / vec;
    for(c = 0; c < n_chunks; ++c)
    {
        int dd = c * vec;
        rocke_value_t* src_addr = rocke_b_add(b, src_base, rocke_b_const_i32(b, dd));
        rocke_value_t* dst_addr = rocke_b_add(b, dst_base, rocke_b_const_i32(b, dd));
        rocke_value_t* v = rocke_b_global_load_vN(b, src, src_addr, ty, vec, vec * elem_bytes);
        rocke_b_global_store_vN(b, dst, dst_addr, v, vec, vec * elem_bytes);
    }
    /* Python: for d in range(n_chunks * vec, H):
     *             s = load_scalar_as_f32(b, src, b.add(src_base, b.const_i32(d)),
     *                                    dtype=dtype)
     *             store_scalar_from_f32(b, dst, b.add(dst_base, b.const_i32(d)),
     *                                   s, dtype=dtype) */
    for(d = n_chunks * vec; d < H; ++d)
    {
        rocke_value_t* s = rocke_b_load_scalar_as_f32(
            b, src, rocke_b_add(b, src_base, rocke_b_const_i32(b, d)), dtype);
        rocke_b_store_scalar_from_f32(
            b, dst, rocke_b_add(b, dst_base, rocke_b_const_i32(b, d)), s, dtype);
    }
}
